src/Elastic.Apm.AspNetFullFramework/Extensions/SoapRequest.cs

```diff
@@ -82,19 +74,50 @@ private static bool IsSoap12Action(NameValueCollection headers)
 			return contentType != null && contentType.Contains(SoapAction12ContentType);
 		}
 
-		internal static string GetSoap12ActionFromInputStream(Stream stream)
+		internal static string GetSoap12ActionFromHttpRequest(IApmLogger logger, HttpRequest request)
 		{
-			StreamReader streamReader = null;
+			//
+			// If the input stream has already been read bufferless, we can't inspect it
+			//
+			if (request.ReadEntityBodyMode == ReadEntityBodyMode.Bufferless)
+			{
+				logger.Debug()?.Log("Request.ReadEntityBodyMode is 'Bufferless'");
+				return null;
+			}
+			//
+			// SOAP 1.2 action must be read from body data.
+			// Using the buffered input stream can have side effects with application code that
+			// runs before or after this.
+			// To be safe, we only access the stream directly if we have random access to it.
+			//
+			var stream = request.InputStream;
+			if (!stream.CanSeek)
+			{
+				logger.Debug()?.Log("Request.InputStream is not seekable");
+				return null;
+			}
+
+			return GetSoap12ActionFromInputStream(logger, stream);
+		}
+
+		internal static string GetSoap12ActionFromInputStream(IApmLogger logger, Stream stream)
+		{ 
+			var shared = ArrayPool<byte>.Shared;
+			var bytes = shared.Rent((int)stream.Length);
 			try
 			{
-				streamReader = new StreamReader(stream);
+				stream.Position = 0;
+				stream.Read(bytes, 0, bytes.Length);
+				stream.Position = 0;
+				var content = Encoding.ASCII.GetString(bytes);
```

Comment:
Seems to be unused. 

```suggestion
```

Reply:
Thanks, a leftover from my dozens of different experiments 😄 .